### ISSUE TITLE
Can I repair a repository with missing segments without using `borg check`?

### ISSUE THREAD

**Issue by finefoot** [borgbackup/borg#4704]
Due to connection issues, some files might not get transferred to the remote server and subsequently are missing from the repository's `data` folder. (This has nothing to do with borg. I'm trying to improve the situation, but for now I have to assume that missing files in `data` might happen. I can explain this further if necessary, but I don't think it's relevant.)

The issue can be reproduced locally with the following steps:

- Create a append-only repository (`append_only = 1`)
- Set `max_segment_size = 1000` to make it easier to produce a lot of segments without taking up a lot of storage
- `borg create repo::one source` which creates `data/0/0` up to `5` for example
- Add some files to the local backup source folder
- `borg create repo::two source` which creates `data/0/6` up to `11` for example
- Finally, manually delete a data segment from `two` like `data/0/7` for example

Now, I can still access all files from `repo::one` without issues. However, naturally, I get errors accessing files from `repo::two` due to the deleted segment `7`.

I don't even need to repair archive `two` - I just want to repair the repo so I can create new archives without referring to the corrupted one. In other words, I *know*  when an archive got corrupted and I want to ignore all changes from that archive as they might refer to segments which don't exist.

What I tried to do:

- If I run `borg delete --force --force repo::two`, the corrupted archive `two` gets deleted.
- However, a subsequent call to `borg create repo::three` as a second try to create a new archive still tries to access missing segment `7` from the now deleted archive `two` and I get `FileNotFoundError`.

Can I make the repo "forget" that archive `two` was ever created?

**Comment by ThomasWaldmann** [comment]
You shouldn't use tools that silently corrupt the borg repo while transferring it. In case of the usual syncing tools, one approach might be to repeatedly run the tool until it terminates without an error.

About the double-force delete:
doesn't it tell you that you shall run `borg check --repair` afterwards to "cleanup the mess"?

An alternative way would be to roll back the transaction that was corrupted by your transfer tool via borg's transaction log, but there is no automation for that, see the docs about how to do it manually. Of course this is only advisable if the corruption is within one of the last repo transaction as you'ld lose all stuff from the rolled back transactions.

**Comment by finefoot** [comment]
> You shouldn't use tools that silently corrupt the borg repo while transferring it. In case of the usual syncing tools, one approach might be to repeatedly run the tool until it terminates without an error.

Yes, I agree.

> About the double-force delete: doesn't it tell you that you shall run borg check --repair afterwards to "cleanup the mess"?

No, it did not. Prompted for passphrase then returned to the shell. No further infos or warnings.

> An alternative way would be to roll back the transaction that was corrupted by your transfer tool via borg's transaction log, but there is no automation for that, see the docs about how to do it manually.

You're referring to https://borgbackup.readthedocs.io/en/stable/usage/notes.html#append-only-mode right? I will still have to rebuild the cache completely from scratch, did I get that right?

> Of course this is only advisable if the corruption is within one of the last repo transactions as you'ld lose all stuff from the rolled back transactions.

Yes, that makes sense. But it's not an issue in this case because I can check immediately after finishing `borg create` if all files have been transferred successfully.

**Comment by ThomasWaldmann** [comment]
Yes, that is what I was referring to. Cache is also described there.

**Comment by finefoot** [comment]
> > About the double-force delete: doesn't it tell you that you shall run borg check --repair afterwards to "cleanup the mess"?
>
> No, it did not. Prompted for passphrase then returned to the shell. No further infos or warnings. (This is with version 1.1.8 - I'll update and report back.)

Just letting you know - same behavior for 1.1.10:

```
$ ./borg-linux64 --version
borg-linux64 1.1.10
$ ./borg-linux64 init --encryption=repokey repo2

By default repositories initialized with this version will produce security
errors if written to with an older version (up to and including Borg 1.0.8).

If you want to use these older versions, you can disable the check by running:
borg upgrade --disable-tam repo2

See https://borgbackup.readthedocs.io/en/stable/changes.html#pre-1-0-9-manifest-spoofing-vulnerability for details about the security implications.

IMPORTANT: you will need both KEY AND PASSPHRASE to access this repo!
Use "borg key export" to export the key, optionally in printable format.
Write down the passphrase. Store both at safe place(s).

$ ./borg-linux64 config repo2 append_only 1
$ ./borg-linux64 config repo2 max_segment_size 1000
$ ./borg-linux64 create repo2::one local
$ ./borg-linux64 create repo2::two local
$ ./borg-linux64 list repo2
one                                  Fri, 2019-08-02 15:41:07 [################################################################]
two                                  Fri, 2019-08-02 15:41:23 [################################################################]
$ rm repo2/data/0/7
$ ./borg-linux64 delete --force --force repo2::two
$ ./borg-linux64 list repo2
one                                  Fri, 2019-08-02 15:41:07 [################################################################]
$ 
```

**Comment by finefoot** [comment]
> Yes, that is what I was referring to. Cache is also described there.

Thanks for confirming. I had read this but wanted to make sure.

Also, it works, by the way. :+1: It makes sense that it's possible to interpret the corrupted files as an malicious attack and therefore just use that recovery procedure. Unfortunately, it's just not feasible to rebuild the cache each time. :( Is it a necessity to have to rebuild the cache or is there just (currently) no way to rollback the cache, too?

Additionally, are there intentions to implement something like `borg rollback` to automatize the mentioned steps in https://borgbackup.readthedocs.io/en/stable/usage/notes.html#append-only-mode?

**Comment by ThomasWaldmann** [comment]
You did not see the message because it was emitted on INFO level, but if you do not use -v, the default logging is at >= WARNING.

I changed the message to WARNING in PR #4705, thanks for finding that.

**Comment by ThomasWaldmann** [comment]
There is no easy way to rollback the cache for a backup that was completed successfully (considering borg). We have locally cached single-archive caches that can be used to compute the current "overall chunks index" (maybe check if they are used or if all is deleted and re-fetched from remote).

**Comment by ThomasWaldmann** [comment]
I don't intend to automate repo roll-back (at least not in the foreseeable future), there are tons of more important things to implement / refactor.

**Comment by finefoot** [comment]
Sorry for taking so long to respond.

> We have locally cached single-archive caches that can be used to compute the current "overall chunks index" (maybe check if they are used or if all is deleted and re-fetched from remote).

Can you elaborate on that? Isn't there a local cache in `.cache/borg` per *repository*, not per single archive? Check if all is deleted? Why would it be deleted? Re-fetch from remote?! :thinking: I'm confused - did I miss a borg feature somehow?

> I don't intend to automate repo roll-back (at least not in the foreseeable future), there are tons of more important things to implement / refactor.

I understand. However, it would be a useful feature, no?

- If the repo is stored remotely, any transmission error (and subsequent defective files) affects the whole repo - even if the error is from a 1 MB addition to a 10 TB repo. `borg check` takes unnecessarily long - a rollback feature would be nice to have.
- Append-only mode for repos is a promoted feature of borg. So if there actually has been malicious modification to an append-only repo by an attacker (which is already a bad situation), there should also be a user-friendly way to make use of the promoted feature, i. e. to comfortably roll back the repo.

Just asking. :)

**Comment by ThomasWaldmann** [comment]
Inside the per-repo cache directory, there is a `chunks.archive.d` subdirectory with per-archives chunks indexes. They can be used to locally create a new master chunks index without fetching all the chunks infos from remote.

**Comment by dragetd** [comment]
This is an old thread, sorry for digging this out @ThomasWaldmann. But I just ran into the same problem. I used rclone mount to use cloud-storage for backups. It is only a testing-setup and I know how flaky this is.
But during a connection issue I am now left with an archive missing one or more segment files. (just test-archive)

I know that we build on the assumption 'we trust the storage'. You have said yourself clearly that there is no priority in implementing forward-error-correction schemes like par2 (#225) or this issue above.
But I think the reality is, that storage is not always perfect. Even without crazy non-intended cloud-fuse setups.

The overall robustness perception of Borg might improve, if we change our assumption here. The issue about FEC is also still open https://github.com/borgbackup/borg/issues/225.
So how about also having an issue to track ideas on how to recover from missing segment files like the idea you mentioned above? (I can open a new issue if you want)

Users in that situation might very well get a clear 'Your storage system fucked up' message, but can at least recover and earlier state of the repository.
